### PR TITLE
Replace tags in email body with unsubscribe and preference links

### PR DESCRIPTION
**Pull request by VenkataNeeharikaKondipati**
https://segment.atlassian.net/browse/CONMAN-443
- Add logic logic to update the unsubscribe tag with links from the externalId payload in sendgrid destination
- Add UTs

_A summary of your pull request, including the what change you're making and why._

## Testing

- Added [unit tests]
- Tested end-to-end functionality in staging and recorded a video: https://twilio.slack.com/archives/C044D7CFYUC/p1687478323951159
- Tested in the staging environment using traditional [deploy](https://github.com/segmentio/eks-configuration/pull/4935)

### CODE REVIEW

**Review by joe-ayoub-segment — COMMENTED** [review]
Looks OK to me @VenkataNeeharikaKondipati . 
What sort of testing has been done on this change? Are you able to share any detail?

FYI We're aiming for a Wednesday deploy next week (due to Tuesday being 04-July). 

cc : @varadarajan-tw

**Review by sumitbhanwala — APPROVED** [review]
Looks good to me.

File: packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts [hunk 1]
```diff
@@ -26,6 +26,42 @@ const insertEmailPreviewText = (html: string, previewText: string): string => {
...
+  if (groupId && groupId != '') {
+    const group = emailProfile?.groups.find((group: { id: string }) => group?.id === groupId)
+    const groupUnsubscribeLink = group?.groupUnsubscribeLink
+    updatedHtml = html.replace(`<a href="${unsubscribeLinkTag}">`, function () {
+      logger?.info(`TE Messaging: Email Group Unsubscribe link replaced  group - ${groupId} spaceId - ${spaceId}`)
+      statsClient?.incr('actions-personas-messaging-sendgrid.replaced_group_unsubscribe_link', 1, tags)
+      return `<a href="${groupUnsubscribeLink}">`
+    })
+  } else {
+    updatedHtml = html.replace(`<a href="${unsubscribeLinkTag}">`, function () {
```
[the remark this answers is not in the document]
Comment (VenkataNeeharikaKondipati) [reply packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts:52]: Good point. Yes, that will work (It was also my initial implementation) But the draw back for that is, if this unsubscribeLinkTag appears in the body without a html link tag - we would replace that as well which is not expected (Though i believe its an edge case to have this tag in the body) I will explore using regex

**Comment by VenkataNeeharikaKondipati** [comment]
> Looks OK to me @VenkataNeeharikaKondipati . What sort of testing has been done on this change? Are you able to share any detail?
> 
> FYI We're aiming for a Wednesday deploy next week (due to Tuesday being 04-July).
> 
> cc : @varadarajan-tw

Added details about testing. Tested changes in stage by deploying actions-destinations traditionally as fast deploy was not picking up the new fields

File: packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts [hunk 2]
```diff
@@ -26,6 +26,56 @@ const insertEmailPreviewText = (html: string, previewText: string): string => {
...
+    $(unsubscribeLinkRef).each(function () {
+      logger?.info(`TE Messaging: Email Global Unsubscribe link replaced  - ${spaceId}`)
+      statsClient?.incr('actions-personas-messaging-sendgrid.replaced_global_unsubscribe_link', 1, tags)
+      $(this).attr('href', globalUnsubscribeLink)
+    })
+  }
+  $(preferencesLinkRef).each(function () {
+    logger?.info(`TE Messaging: Email Preferences link replaced  - ${spaceId}`)
+    statsClient?.incr('actions-personas-messaging-sendgrid.replaced_preferences_link', 1, tags)
+    if (!preferencesLink || preferencesLink == '') {
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts:61]: Empty string is a falsy value. So, it should be okay to do just do the below here.
```suggestion
    if (!preferencesLink) {
```

File: packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts [hunk 3]
```diff
@@ -26,6 +26,56 @@ const insertEmailPreviewText = (html: string, previewText: string): string => {
...
+  tags: string[],
+  groupId?: string,
+  logger?: Logger | undefined
+): string => {
+  const globalUnsubscribeLink = emailProfile?.unsubscribeLink
+  const preferencesLink = emailProfile?.preferencesLink
+  const unsubscribeLinkRef = 'a[href*="[upa_unsubscribe_link]"]'
+  const preferencesLinkRef = 'a[href*="[upa_preferences_link]"]'
+  const $ = cheerio.load(html)
+  if (groupId && groupId != '') {
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/engage-messaging-sendgrid/sendEmail/index.ts:43]: Empty string is a falsy value. So, it should be okay to check `if(groupId)`

**Review by varadarajan-tw — APPROVED** [review]
Minor nits b

File: packages/destination-actions/src/destinations/engage-messaging-sendgrid/__tests__/send-email.test.ts [hunk 4]
```diff
@@ -692,6 +698,258 @@ describe.each([
...
+            ],
+            bcc: [
+              {
+                email: 'test@test.com'
+              }
+            ],
+            custom_args: {
+              source_id: 'sourceId',
+              space_id: 'spaceId',
+              user_id: userData.userId,
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/engage-messaging-sendgrid/__tests__/send-email.test.ts:807]: I just merged a [PR](https://github.com/segmentio/action-destinations/commit/3c272ef41f6ecf117f9949f839cffa025440d742) that removes user_id from custom_args. You might want to rebase/merge master to your branch.

**Review by sumitbhanwala — APPROVED** [review]
LGTM

**Comment by varadarajan-tw** [comment]
Hey  @VenkataNeeharikaKondipati, this PR has been deployed on Wed 5, July!

**Pull request merged**